### PR TITLE
[ISSUE-636][FEATURE] worker sent user usage to master through register worker/heartbeat from worker and master update usage

### PR DESCRIPTION
**Pull request by nafiyAix**
### What changes were proposed in this pull request?
Send user usage in RPC RegisterWorker & HeartbeatFromWorker to Master  and Master update usage

### Why are the changes needed?
Collect user resource usage from worker to master

### Related issues.
#636 

### How was this patch tested?

/cc @AngersZhuuuu 

/assign @main-reviewer

### CODE REVIEW

File: common/src/main/scala/com/aliyun/emr/rss/common/util/Utils.scala [hunk 1]
```diff
@@ -562,6 +562,17 @@ object Utils extends Logging {
     s"$shuffleId-$mapId-$attemptId"
   }
 
+  def makeUserResourceUsageKey(userIdentifier: UserIdentifier): String = {
+    s"${userIdentifier.tenantId}-${userIdentifier.name}"
+  }
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/util/Utils.scala:567]: Move to `UserIdentifier.toString`?

File: common/src/main/scala/com/aliyun/emr/rss/common/util/Utils.scala [hunk 2]
```diff
@@ -562,6 +562,17 @@ object Utils extends Logging {
...
+
+  def splitUserResourceUsageKey(userResourceUsageKey: String): UserIdentifier = {
+    val splits = userResourceUsageKey.split("-")
+    val tenant = splits.dropRight(1).mkString("-")
+    val username = splits.last
+    UserIdentifier(tenant, username)
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/util/Utils.scala:573]:
```
object UserIdentifier {
   def  apply(user: String) = {
   // Can use Regex like https://github.com/alibaba/RemoteShuffleService/issues/656
   }
}
```

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala [hunk 3]
```diff
@@ -80,15 +80,21 @@ object ControlMessages extends Logging {
       fetchPort: Int,
       replicatePort: Int,
       disks: Map[String, DiskInfo],
+      userResourceUsage: Map[UserIdentifier, ResourceConsumption],
       requestId: String): PbRegisterWorker = {
     val pbDisks = disks.mapValues(PbSerDeUtils.toPbDiskInfo).asJava
+    val pbUserResourceUsage = userResourceUsage
+      .map(x => (Utils.makeUserResourceUsageKey(x._1), x._2))
+      .mapValues(PbSerDeUtils.toPbResourceConsumption)
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:88]: After change `UserIdentifier` serde method, here can be 
```
val pbUserResourceUsage = userResourceUsage
      .map { case (user, resources) =>
           UserIdentifier(user) -> PbSerDeUtils.toPbResourceConsumption(resources)
     }.toMap 
....
```

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala [hunk 4]
```diff
@@ -348,23 +355,21 @@ object ControlMessages extends Logging {
...
-            .newBuilder()
-            .setUsableSpace(item._2.actualUsableSpace)
-            .setAvgFlushTime(item._2.avgFlushTime)
-            .setUsedSlots(item._2.activeSlots)
-            .setStatus(item._2.status.getValue)
-            .build()).toMap.asJava
+      val pbDisks = disks.asScala.mapValues(PbSerDeUtils.toPbDiskInfo).asJava
+      val pbUserResourceUsage = userResourceUsage
+        .map(x => (Utils.makeUserResourceUsageKey(x._1), x._2))
+        .mapValues(PbSerDeUtils.toPbResourceConsumption)
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:364]: ditto

File: server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala [hunk 5]
```diff
@@ -256,6 +262,7 @@ private[deploy] class Master(
           fetchPort,
           replicatePort,
           disks,
+          userResourceUsage,
```
Comment (AngersZhuuuu) [inline server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala:265]: Miss change `handleHeartbeatFromWorker` ?

File: server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala [hunk 6]
```diff
@@ -210,6 +210,12 @@ private[deploy] class Master(
       val fetchPort = pbRegisterWorker.getFetchPort
       val replicatePort = pbRegisterWorker.getReplicatePort
       val disks = pbRegisterWorker.getDisksMap.asScala.mapValues(PbSerDeUtils.fromPbDiskInfo).asJava
+      val userResourceUsage = pbRegisterWorker
+        .getUserResourceUsageMap
+        .asScala
+        .map(x => (Utils.splitUserResourceUsageKey(x._1), x._2))
+        .mapValues(PbSerDeUtils.fromPbResourceConsumption)
+        .asJava
```
Comment (AngersZhuuuu) [inline server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala:218]: miss change `handleRegisterWorker`?

File: common/src/main/proto/TransportMessages.proto [hunk 7]
```diff
@@ -126,7 +126,8 @@ message PbRegisterWorker {
   int32 fetchPort = 4;
   int32 replicatePort = 5;
   map<string, PbDiskInfo> disks = 6;
-  string requestId = 9;
+  string requestId = 7;
+  map<string, PbResourceConsumption> userResourceUsage = 8;
```
Comment (FMX) [inline common/src/main/proto/TransportMessages.proto:130]: Register worker means that the worker is just initialized and empty. So I think there is no need to report empty userResourceUsage unless worker recovering saved that information.

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala [hunk 8]
```diff
@@ -80,15 +80,21 @@ object ControlMessages extends Logging {
       fetchPort: Int,
       replicatePort: Int,
       disks: Map[String, DiskInfo],
+      userResourceUsage: Map[UserIdentifier, ResourceConsumption],
```
Comment (FMX) [inline common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:83]: As I have mentioned before, is it necessary?

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala [hunk 9]
```diff
@@ -335,6 +335,7 @@ private[deploy] class Worker(
               fetchPort,
               replicatePort,
               workerInfo.diskInfos.asScala.toMap,
+              workerInfo.userResourceUsage.asScala.toMap,
```
Comment (FMX) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala:338]: ditto

File: common/src/main/proto/TransportMessages.proto (hunk 7 above)
Comment (AngersZhuuuu) [reply common/src/main/proto/TransportMessages.proto:130]: > Register worker means that the worker is just initialized and empty. So I think there is no need to report empty userResourceUsage unless worker recovering saved that information.

But if it's recovered from graceful shutdown?

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala (hunk 8 above)
Comment (nafiyAix) [reply common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:83]: We plan to achieve recover user usage from levelDB in #642 , so if there is recovered user data, I think we may need to bring this information when registering worker.

File: common/src/main/proto/TransportMessages.proto (hunk 7 above)
Comment (FMX) [reply common/src/main/proto/TransportMessages.proto:130]: OK, it's reasonable.

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala (hunk 8 above)
Comment (FMX) [reply common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:83]: OK.

**Comment by FMX** [comment]
@nafiyAix Hi, if a comment is addressed, resolve it. So I can tell there is no further discussion.

**Comment by nafiyAix** [comment]
@FMX  FYI: Parts of code change is about how Master and IMetadataHandler's related class to handle the userResourceUsage inside `handleRegisterWorker` and `handleHeartbeatFromWorker`.

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala [hunk 10]
```diff
@@ -80,15 +81,21 @@ object ControlMessages extends Logging {
...
       requestId: String): PbRegisterWorker = {
     val pbDisks = disks.mapValues(PbSerDeUtils.toPbDiskInfo).asJava
+    val pbUserResourceUsage = userResourceUsage
+      .map(usage => (usage._1.toString, usage._2))
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:88]: Use
```
map { case (a, b) =>
}
```

File: common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala [hunk 11]
```diff
@@ -371,23 +379,22 @@ object ControlMessages extends Logging {
...
-            .newBuilder()
-            .setUsableSpace(item._2.actualUsableSpace)
-            .setAvgFlushTime(item._2.avgFlushTime)
-            .setUsedSlots(item._2.activeSlots)
-            .setStatus(item._2.status.getValue)
-            .build()).toMap.asJava
+      val pbDisks = disks.asScala.mapValues(PbSerDeUtils.toPbDiskInfo).asJava
+      val pbUserResourceUsage = userResourceUsage
+        .asScala
+        .map(usage => (usage._1.toString, usage._2))
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/protocol/message/ControlMessages.scala:388]: ditto

File: server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala [hunk 12]
```diff
@@ -213,6 +213,12 @@ private[deploy] class Master(
       val fetchPort = pbRegisterWorker.getFetchPort
       val replicatePort = pbRegisterWorker.getReplicatePort
       val disks = pbRegisterWorker.getDisksMap.asScala.mapValues(PbSerDeUtils.fromPbDiskInfo).asJava
+      val userResourceUsage = pbRegisterWorker
+        .getUserResourceUsageMap
+        .asScala
+        .map(x => (UserIdentifier(x._1), x._2))
```
Comment (AngersZhuuuu) [inline server-master/src/main/scala/com/aliyun/emr/rss/service/deploy/master/Master.scala:219]: ditto

File: common/src/main/scala/com/aliyun/emr/rss/common/meta/WorkerInfo.scala [hunk 13]
```diff
@@ -37,6 +39,7 @@ class WorkerInfo(
     var endpoint: RpcEndpointRef) extends Serializable with Logging {
   var unknownDiskSlots = new java.util.HashMap[String, Integer]()
   var lastHeartbeat: Long = 0
+  val userResourceUsage = new ConcurrentHashMap[UserIdentifier, ResourceConsumption]()
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/meta/WorkerInfo.scala:42]: userResourceConsumption?

File: common/src/main/scala/com/aliyun/emr/rss/common/meta/WorkerInfo.scala [hunk 14]
```diff
@@ -213,6 +216,10 @@ class WorkerInfo(
     }
   }
 
+  def updateUserResourceUsage(usage: util.Map[UserIdentifier, ResourceConsumption]): Unit = {
+    userResourceUsage.putAll(usage)
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/meta/WorkerInfo.scala:220]: Should we clean the ResourcesConsumption that not existed in the usage?

**Comment by waitinfuture** [comment]
Since @pan3793 is removing all control messages and unify to pb https://github.com/alibaba/RemoteShuffleService/issues/673 , should we delete ResourceConsumption and merely use pb? Maybe we can do it in later prs.

File: common/src/main/proto/TransportMessages.proto [hunk 15]
```diff
@@ -126,7 +126,8 @@ message PbRegisterWorker {
   int32 fetchPort = 4;
   int32 replicatePort = 5;
   map<string, PbDiskInfo> disks = 6;
-  string requestId = 9;
+  string requestId = 7;
```
[the remark this answers is not in the document]
Comment (nafiyAix) [reply common/src/main/proto/TransportMessages.proto:129]: will rollback it

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**